server/embedded/src/org/labkey/embedded/LabKeyServer.java

```diff
@@ -0,0 +1,358 @@
+package org.labkey.embedded;
+
+import org.apache.catalina.core.StandardContext;
+import org.apache.catalina.loader.WebappLoader;
+import org.apache.catalina.startup.Tomcat;
+import org.apache.tomcat.util.descriptor.web.ContextResource;
+import org.labkey.bootstrap.ConfigException;
+import org.springframework.boot.SpringApplication;
+import org.springframework.boot.autoconfigure.SpringBootApplication;
+import org.springframework.boot.context.properties.ConfigurationProperties;
+import org.springframework.boot.web.embedded.tomcat.TomcatServletWebServerFactory;
+import org.springframework.boot.web.embedded.tomcat.TomcatWebServer;
+import org.springframework.context.ApplicationContext;
+import org.springframework.context.annotation.Bean;
+import org.springframework.context.annotation.Configuration;
+import org.springframework.validation.annotation.Validated;
+
+import javax.sql.DataSource;
+import javax.validation.constraints.NotEmpty;
+import javax.validation.constraints.NotNull;
+import java.io.BufferedOutputStream;
+import java.io.File;
+import java.io.FileOutputStream;
+import java.io.IOException;
+import java.io.InputStream;
+import java.util.ArrayList;
+import java.util.Enumeration;
+import java.util.List;
+import java.util.jar.JarEntry;
+import java.util.jar.JarFile;
+import java.util.zip.ZipEntry;
+import java.util.zip.ZipInputStream;
+
+@SpringBootApplication
+public class LabKeyServer
+{
+	private static final int BUFFER_SIZE = 4096;
+
+	public static void main(String[] args)
+	{
+		ApplicationContext ctx = SpringApplication.run(LabKeyServer.class, args);
+	}
+
+	@Bean
+	public ContextProperties labkeyDataSource()
+	{
+		return new ContextProperties();
+	}
+
+	@Bean
+	public MailProperties smtpSource()
+	{
+		return new MailProperties();
+	}
+
+	@Bean
+	public TomcatServletWebServerFactory servletContainerFactory()
+	{
+		return new TomcatServletWebServerFactory()
+		{
+			@Override
+			protected TomcatWebServer getTomcatWebServer(Tomcat tomcat)
+			{
+				tomcat.enableNaming();
+
+				// Get the datasource properties from Spring injection
+				ContextProperties props = labkeyDataSource();
+
+				// for development, point to the local deploy/labkeyWebapp directory
+				boolean webAppLocationPresent = props.getWebAppLocation() != null;
+				var webAppLocation = "";
+
+				if (!webAppLocationPresent)
+				{
+					var destDirectory = new File("").getAbsolutePath() + "/server";
+					webAppLocation = destDirectory + "/labkeywebapp";
+					boolean extracted = new File(webAppLocation).exists();
```

Comment:
Let's talk this through. It should be OK to have multiple JARs for different versions in the same directory, and be able to start with any of them and have it do the right thing.